common/unsafe/src/main/java/org/apache/spark/unsafe/memory/UnsafeMemoryAllocator.java

```diff
@@ -36,7 +37,7 @@ public MemoryBlock allocate(long size) throws OutOfMemoryError {
 
   @Override
   public void free(MemoryBlock memory) {
-    assert (memory.obj == null) :
+    assert (memory.obj instanceof MemoryAddress) :
       "baseObject not null; are you trying to use the off-heap allocator to free on-heap memory?";
```

Comment:
Would having a null test still be appropriate? So adding it as a condition to the `instanceof`